drv/cosmo-hf/src/main.rs

```diff
@@ -0,0 +1,294 @@
+    pub const FAST_READ_QUAD: u32 = 0x6b;
+    pub const FAST_READ_QUAD_OUTPUT_4B: u32 = 0x6c;
+    pub const SECTOR_ERASE: u32 = 0x20;
+    pub const READ_JEDEC_ID: u32 = 0x9f;
+    pub const BLOCK_ERASE_64KB: u32 = 0xd8;
+    pub const BLOCK_ERASE_64KB_4B: u32 = 0xdc;
+    pub const QUAD_INPUT_PAGE_PROGRAM: u32 = 0x32;
+    pub const QUAD_INPUT_PAGE_PROGRAM_4B: u32 = 0x34;
+}
+
+impl ServerImpl {
+    fn flash_read_id(&self) -> [u8; 20] {
+        self.clear_fifos();
+        self.write_reg(reg::DATA_BYTES, 20);
+        self.write_reg(reg::ADDR, 0);
+        self.write_reg(reg::DUMMY_CYCLES, 0);
+        self.write_reg(reg::INSTR, instr::READ_JEDEC_ID);
+        self.wait_fpga_busy();
+        let mut out = [0u8; 20];
+        for i in 0..out.len() / 4 {
+            let v = self.read_reg(reg::RX_FIFO);
+            for (j, byte) in v.to_le_bytes().iter().enumerate() {
+                out[i * 4 + j] = *byte;
+            }
+        }
+        out
+    }
+
+    fn read_reg(&self, reg: *mut u32) -> u32 {
+        unsafe { reg.read_volatile() }
+    }
+
+    fn write_reg(&self, reg: *mut u32, v: u32) {
+        unsafe { reg.write_volatile(v) };
+    }
+
+    /// Wait until the FPGA is idle
+    fn wait_fpga_busy(&self) {
+        loop {
+            let status = self.read_reg(reg::SPISR);
+            if (status & 1) == 0 {
+                break;
+            }
+            ringbuf_entry!(Trace::FpgaBusy(status));
+            sleep_for(1);
+        }
+    }
+
+    /// Wait until a word is available in the RX buffer
+    fn wait_fpga_rx(&self) {
+        for i in 0.. {
+            let status = self.read_reg(reg::SPISR);
+            // Wait for the
+            const RX_EMPTY_BIT: u32 = 1 << 6;
+            if status & RX_EMPTY_BIT == 0 {
+                break;
+            }
+            ringbuf_entry!(Trace::FpgaBusy(status));
+            // Initial busy-loop for faster response
+            if i >= 32 {
+                sleep_for(1);
+            }
+        }
+    }
+
+    /// Wait until the SPI flash is idle
+    fn wait_flash_busy(&self, t: Trace) {
+        // Wait for the busy flag to be unset
+        while (self.read_flash_status() & 1) != 0 {
+            ringbuf_entry!(t);
+            sleep_for(1);
+        }
+    }
+
+    fn read_flash_status(&self) -> u8 {
+        self.clear_fifos();
+        self.write_reg(reg::DATA_BYTES, 1);
+        self.write_reg(reg::ADDR, 0);
+        self.write_reg(reg::DUMMY_CYCLES, 0);
+        self.write_reg(reg::INSTR, instr::READ_STATUS_1);
+        self.wait_fpga_busy();
+        self.read_reg(reg::RX_FIFO).to_le_bytes()[0]
+    }
+
+    fn flash_write_enable(&self) {
+        self.write_reg(reg::DATA_BYTES, 0);
+        self.write_reg(reg::ADDR, 0);
+        self.write_reg(reg::DUMMY_CYCLES, 0);
+        self.write_reg(reg::INSTR, instr::WRITE_ENABLE);
+        self.wait_fpga_busy();
+    }
+
+    fn clear_fifos(&self) {
+        self.write_reg(reg::SPICR, 0x8080);
```

Comment:
**question:** Other operations generally need to wait for the FPGA to become idle after acting; that's presumably not the case here?

Reply:
There is no wait required here. The FPGA FIFO clears take single-digit clock cycles to complete, as the reset may have to cross clock domains to clear read/write pointers, but the SP is hanging off the FMC interface and can't even finish executing another read/write transaction to the FPGA before the FIFO clear has completed so this can be treated as if it happens instantly.